include/aws/common/zero.h

```diff
@@ -0,0 +1,88 @@
+#ifndef AWS_COMMON_ZERO_H
+#define AWS_COMMON_ZERO_H
+
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/stdbool.h>
+#include <aws/common/stdint.h>
+
+#include <string.h>
+
+/**
+ * Set each byte in the struct to zero.
+ */
+#define AWS_ZERO_STRUCT(object)                                                                                        \
+    do {                                                                                                               \
+        memset(&(object), 0, sizeof(object));                                                                          \
+    } while (0)
+
+/**
+ * Set each byte in the array to zero.
+ */
+#define AWS_ZERO_ARRAY(array)                                                                                          \
+    do {                                                                                                               \
+        memset((void *)array, 0, sizeof(array));                                                                       \
+    } while (0)
+
+/**
+ * Returns whether each byte in the object is zero.
+ */
+#define AWS_IS_ZEROED(object) aws_is_zeroed(&(object), sizeof(object))
+
+/**
+ * Returns whether each byte is zero.
+ */
+AWS_STATIC_IMPL
+bool aws_is_zeroed(const void *buf, size_t bufsize) {
+    /* Optimization idea: vectorized instructions to check more than 64 bits at a time. */
+
+    /* Check 64 bits at a time */
+    const uint64_t *buf_u64 = (const uint64_t *)buf;
+    const size_t num_u64_checks = bufsize / 8;
+    size_t i;
+    for (i = 0; i < num_u64_checks; ++i) {
+        if (buf_u64[i]) {
+            return false;
+        }
+    }
+
+    /* Update buf to where u64 checks left off */
+    buf = buf_u64 + num_u64_checks;
+    bufsize = bufsize % 8;
+
+    /* Check 8 bits at a time */
+    const uint8_t *buf_u8 = (const uint8_t *)buf;
```

Comment:
You can replace this loop with a ```switch(bufsize & 0x7) { ...``` unrolling.

Reply:
I ran this through Godbolt and the compiler did unroll it
I didn't want to get *too* hung up on theoretical efficiency improvements, since the obvious one is to use vectorized instructions, but I don't want to spend a week getting that working across all compilers right now. I just want to get a good-enough version in.

Reply:
https://godbolt.org/z/5UG8je

Reply:
I also thought about a different version for 32bit processors that took 32bit steps instead of 64bit steps, but it compiled down to essentially the same thing, so I just kept it simple.